ruby/ext/google/protobuf_c/encode_decode.c

```diff
@@ -174,9 +174,222 @@ static void *submsg_handler(void *closure, const void *hd) {
   return submsg;
 }
 
+// Handler data for startmap/endmap handlers.
+typedef struct {
+  size_t ofs;
+  const upb_fielddef* key_field;
+  const upb_fielddef* value_field;
+  VALUE value_field_typeclass;
+} map_handlerdata_t;
+
+// Temporary frame for map parsing: at the beginning of a map entry message, a
+// submsg handler allocates a frame to hold (i) a reference to the Map object
+// into which this message will be inserted and (ii) storage slots to
+// temporarily hold the key and value for this map entry until the end of the
+// submessage. When the submessage ends, another handler is called to insert the
+// value into the map.
+typedef struct {
+  VALUE map;
+  char key_storage[NATIVE_SLOT_MAX_SIZE];
+  char value_storage[NATIVE_SLOT_MAX_SIZE];
+} map_parse_frame_t;
+
+// Handler to begin a sequence of map entries: simple no-op that exists only to
+// set context for the map entry handlers.
+static void *startmap_handler(void *closure, const void *hd) {
```

Comment:
I don't believe this should be necessary; in the absence of a start handler, the outer closure should get propagated.